front/src/HOC/withLogger.test.js

```diff
@@ -0,0 +1,211 @@
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    rerender(<WrappedComponent qw="1" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(3);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(4);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component should update'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      3,
+      'component updated'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      4,
+      'component will unmount'
+    );
+  });
+  test('should add log function with specified HOC name', () => {
+    const WrappedComponent = withLogger()(TestComponent);
+    expect(WrappedComponent.displayName).toBe(
+      'withLogger(TestComponent)'
+    );
+  });
+  test('should log specified lifecycles', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['didMount', 'willUnmount'],
+    })(TestComponent);
+    const { unmount } = render(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(2);
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      1,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component will unmount'
+    );
+  });
+  test('should log props', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: ['props'],
+    })(TestComponent);
+    const {} = render(<WrappedComponent q="1" w="erty" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(1);
+    expect(fakeLogger).toHaveBeenNthCalledWith(1, {
+      q: '1',
+      w: 'erty',
+    });
+  });
+  test('should log props first', () => {
+    const WrappedComponent = withLogger({
+      output: fakeLogger,
+      lifecycles: [
+        'props',
+        'didMount',
+        'shouldUpdate',
+        'willUnmount',
+      ],
+    })(TestComponent);
+    const { rerender, unmount } = render(
+      <WrappedComponent q="1" w="erty" />
+    );
+    expect(fakeLogger).toHaveBeenCalledTimes(2);
+    rerender(<WrappedComponent qw="1" />);
+    expect(fakeLogger).toHaveBeenCalledTimes(4);
+    unmount(<WrappedComponent />);
+    expect(fakeLogger).toHaveBeenCalledTimes(5);
+    expect(fakeLogger).toHaveBeenNthCalledWith(1, {
+      q: '1',
+      w: 'erty',
+    });
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      2,
+      'component mounted'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(3, {
+      qw: '1',
+    });
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      4,
+      'component should update'
+    );
+    expect(fakeLogger).toHaveBeenNthCalledWith(
+      5,
+      'component will unmount'
+    );
+  });
+  test('should add log function with specified params for class component', () => {
```

Comment:
That is an interesting test case. Could you please give some usage example of such a feature or when it is necessary/better than the default approach (above)?

Reply:
This test is for checking the order of parent/child lifecycles.